inform7/Internal/Inter/Architecture32Kit/Sections/Glk.i6t

```diff
@@ -4,1001 +4,502 @@ Glk is a fully-featured input-output model, designed by Andrew Plotkin for use
-  @glk 312 _vararg_count ret;
-  return ret;
-];
+	! Run the glk event handling rules (but disable rules debugging because it crashes if keyboard input events are pending)
+	@push debug_rules; @push say__p; @push say__pc;
+	debug_rules = false; ClearParagraphing(1);
+	FollowRulebook(GLK_EVENT_HANDLING_RB, Glk_Event_Struct_type, true);
+	@pull say__pc; @pull say__p; @pull debug_rules;
 
-[ glk_stream_open_memory_uni _vararg_count ret;
-  ! glk_stream_open_memory_uni(uintarray, arraylen, uint, uint) => stream
-  @glk 313 _vararg_count ret;
-  return ret;
-];
+	! If the current event type is line input, update the current buffer length
+	if (Glk_Event_Struct_type == evtype_LineInput && Glk_Event_Struct_win) {
+		Glk_Event_Struct_val1 = Glk_Event_Struct_win.line_input_buffer_curlen;
+	}
 
-[ glk_stream_open_resource_uni _vararg_count ret;
-  ! glk_stream_open_resource_uni(uint, uint) => stream
-  @glk 314 _vararg_count ret;
-  return ret;
+	! Copy back to the original event structure
+	event_struct-->0 = Glk_Event_Struct_type;
+	if (Glk_Event_Struct_win) {
+		event_struct-->1 = GProperty(OBJECT_TY, Glk_Event_Struct_win, glk_ref);
+	}
+	else {
+		event_struct-->1 = 0;
+	}
+	event_struct-->2 = Glk_Event_Struct_val1;
+	event_struct-->3 = Glk_Event_Struct_val2;
+	rfalse;
 ];
 
-[ glk_request_char_event_uni _vararg_count;
-  ! glk_request_char_event_uni(window)
-  @glk 320 _vararg_count 0;
-  return 0;
-];
+@ |SetGlkEventType| helps us change the current event type, by ensuring that
+if there are any pending keyboard input requests they will be cancelled if the
+new event is a keyboard event. It does not however do anything if no keyboard
+input requests are pending, which means the new event could be ignored.
 
-[ glk_request_line_event_uni _vararg_count;
-  ! glk_request_line_event_uni(window, uintarray, arraylen, uint)
-  @glk 321 _vararg_count 0;
-  return 0;
+=
+[ SetGlkEventType type;
+	if (type == evtype_CharInput or evtype_LineInput) {
+		if (Glk_Event_Struct_win.text_input_status == INPUT_STATUS_ACTIVE_CHAR or INPUT_STATUS_ACTIVE_CHAR_UNI) {
+			glk_cancel_char_event(Glk_Event_Struct_win.glk_ref);
+		}
+		else if (Glk_Event_Struct_win.text_input_status == INPUT_STATUS_ACTIVE_LINE or INPUT_STATUS_ACTIVE_LINE_UNI) {
+			glk_cancel_line_event(Glk_Event_Struct_win.glk_ref);
+		}
+		Glk_Event_Struct_win.text_input_status = INPUT_STATUS_NONE;
+	}
+	Glk_Event_Struct_type = type;
 ];
 
-[ glk_set_echo_line_event _vararg_count;
-  ! glk_set_echo_line_event(window, uint)
-  @glk 336 _vararg_count 0;
-  return 0;
-];
+@h Tracking input requests.
+These replaced versions of the glk versions track which types of input have been
+requested.
 
-[ glk_set_terminators_line_event _vararg_count;
-  ! glk_set_terminators_line_event(window, uintarray, arraylen)
-  @glk 337 _vararg_count 0;
-  return 0;
+=
+[ glk_cancel_char_event win win_obj;
+	@push win;
+	@glk 211 1 0;
+	win_obj = FindGlkWindowFromRefNum(win);
```

Comment:
Just had the thought that, for better performance, I7 could pass in the win_obj to these functions. So it should only search if it's not set.